package/system/procd/files/nand.sh

```diff
@@ -237,6 +248,111 @@ nand_upgrade_ubifs() {
 	nand_do_upgrade_success
 }
 
+# Flash kernel part of combined image to kernel partition
+# and rootfs part(squashfs) to ubi_gluebi rootfs partition
+nand_upgrade_combined() {
+	local CI_BLKSZ=65536
+	local kernelpart=$CI_KERNPART
+	local rootfspart=rootfs
+	local kern_length=0x$(dd if="$1" bs=2 skip=1 count=4 2>/dev/null)
+	local kern_blocks=$(($kern_length / $CI_BLKSZ))
+	local kern_magic_id=$(dd if="$1" bs=1 skip=$CI_BLKSZ count=8 2>/dev/null)
+	local rootfs_length=0x$(dd if="$1" bs=2 skip=5 count=4 2>/dev/null)
+	local root_blocks=$((0x$(dd if="$1" bs=2 skip=5 count=4 2>/dev/null) / $CI_BLKSZ))
+	local rootfs_type=$(identify $1 $((($kern_length + $CI_BLKSZ) / 4)))
+	#write kernel
+	[ "$kern_magic_id" != "MIKROTIK" ] && {
+		dd if="$1" bs=$CI_BLKSZ skip=1 count=$kern_blocks 2>/dev/null | \
+			    mtd write - $kernelpart
+        } #else: MIKROTIK: Here I do not need to write kernel. I allready done this in platform_pre_upgrade.
+	#write rootfs to ubi0->rootfs
+	nand_upgrade_prepare_ubi "$rootfs_length" "$rootfs_type" "0" "0"
+
+	local ubidev="$(nand_find_ubi "$CI_UBIPART")"
+	local root_ubivol="$(nand_find_volume $ubidev rootfs)"
+	dd if="$1" bs=$CI_BLKSZ skip=$((1+$kern_blocks)) count=$root_blocks 2>/dev/null | \
+		ubiupdatevol /dev/$root_ubivol -s $rootfs_length -
+	nand_do_upgrade_success "jffs2"
+}
+
+#Return common sysupgrade platform name(NNDXXXXX or NORXXXXX) for Mikrotik devices
+get_mikrotik_sysupgrade_platform_name(){
+	local kern_mtd_index="$(find_mtd_index $CI_KERNPART)"
+	local kern_mtd_dev="/dev/mtd$kern_mtd_index"
+	[ -n "$kern_mtd_index" -a -e "$kern_mtd_dev" ] || return 1
+	[ -x /usr/sbin/mtdinfo ] || {
+		echo "Error! mtdinfo not found! Package nand-utils does not installed ?"
+		return 1
+	}
+	local mtd_info=$(/usr/sbin/mtdinfo $kern_mtd_dev)
+	local mtd_type=`echo "$mtd_info" | sed -n 's/^Type: *\(.\+\)$/\1/p'`
+	local mtd_unit_size=`echo "$mtd_info" | sed -n 's/^Minimum input\/output unit size\: *\([0-9]\+\) *bytes$/\1/p'`
+	[ -z "$mtd_type" -o -z "$mtd_unit_size" ] && return 1
+	if [ "$mtd_type" == "nand" ]; then
+		mtd_type="NND"
+	elif [ "$mtd_type" == "nor" ]; then
+		mtd_type="NOR"
+	else
+		return 1
+	fi
```

Comment:
Stylistic remark: a `case ... esac` expression would make the code above look better